### ISSUE TITLE
getSession({req}) not working in api route

### ISSUE THREAD

**Issue by louisjrdev** [nextauthjs/next-auth#4620]
### Question 💬

Hello!

I have an api route which I am trying to protect by getting the session using `const session = await getSession({req})`.

However, it always returns `null`; Any help diagnosing the issue here is much appreciated, I have looked through the example app and I really can't see what I am doing differently that would cause this.

### How to reproduce ☕️

My api route is as follows: 

```js
import { getSession } from "next-auth/react";

const { CosmosClient } = require("@azure/cosmos");

export default async (req, res)  => {
    const session = await getSession({req});
    console.log(session); //null
}
```
[...nextauth].js
```js
import NextAuth from "next-auth"
import GithubProvider from "next-auth/providers/github"

export default NextAuth({
  providers: [
    GithubProvider({
      clientId: process.env.GITHUB_ID,
      clientSecret: process.env.GITHUB_SECRET,
    })
  ],
})
```
_app.js

```js
import { SessionProvider } from "next-auth/react"
import { useSession, signIn, signOut } from "next-auth/react"

export default function App({
  Component,
  pageProps: { session, ...pageProps },
}) {
  return (
    <SessionProvider session={session}>
      <User></User>
      <Component {...pageProps} />
    </SessionProvider>
  )
}

function User() {
  const { data: session } = useSession()
  if (session) { 
    return (
      <>        
        Signed in as {session.user.email} <br />        
        <button onClick={() => signOut()}>Sign out</button>      
      </>
    )
  } 
  return (
    <>      
      Not signed in <br />      
      <button onClick={() => signIn()}>Sign in</button>    
    </>
  )
}

```

It is worth noting that the sign in / sign out works correctly and shows my email when I am logged in at the same time that getSession() in my api route returns null

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by Klin-Coders** [comment]
i hope you are not accessing the api using postman or something related ?

**Comment by louisjrdev** [comment]
@Klin-Coders Nope, just a plain ol' fetch in getServerSideProps of one of my pages

**Comment by Klin-Coders** [comment]
firstly its is not advised to call your api routes using getServerSideProps...

but do you mind showing how you are using it in getServerSideProps??
to see if i can help maybe?

**Comment by louisjrdev** [comment]
@Klin-Coders 

> firstly its is not advised to call your api routes using getServerSideProps...
As far as I can see, this is not discouraged, it is a valid use case and NextJS nor Next Auth mention anything to suggest you cannot / should not do data fetching in getServerSideProps.

This is my function
```js
export async function getServerSideProps(ctx){
    let session = await getSession(ctx);
    console.log("Session", session); // Not null here
    let response = await fetch(process.env.NEXT_PUBLIC_HOST+`/api/orders/getAll?page=${ctx.query.page}`);
    if(response.status == 200){
        var blob = await response.blob();
        var json = await blob.text();
        var responseJson = JSON.parse(json);
        return { props: { data: responseJson, session: session } };
    }else{
        return { props: { data: {}, session: session } };
    }
  }
```

**Comment by Klin-Coders** [comment]
I do not see a reason why it should be null, this seems clean
i would log my req object and see if all the necessary credentials are being sent to the api route 
sorry i cannot be of much help, lets wait for the pros

but then here is the link to the NextJS docs that explains why you should not call your api routes from ```getServerSideProps```

[getServerSideProps or API Routes](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#getserversideprops-or-api-routes)

It can be tempting to reach for an [API Route](https://nextjs.org/docs/api-routes/introduction) when you want to fetch data from the server, then call that API route from getServerSideProps. This is an unnecessary and inefficient approach, as it will cause an extra request to be made due to both getServerSideProps and API Routes running on the server.

Take the following example. An API route is used to fetch some data from a CMS. That API route is then called directly from getServerSideProps. This produces an additional call, reducing performance. Instead, directly import the logic used inside your API Route into getServerSideProps. This could mean calling a CMS, database, or other API directly from inside getServerSideProps.

**Comment by louisjrdev** [comment]
@Klin-Coders That is actually helpful, perhaps next-auth just doesn't support getSession() when calling the api route from getServerSideProps.

I have instead moved my CRUD code to a service, for which I can call in getServerSideProps and when I need to make client side calls (useEffect) I still hit an API endpoint, but just re-use the service I wrote.

Thank you! It would still be good to get an answer however, on whether the initial scenario i was trying should be possible.

**Comment by Klin-Coders** [comment]
yes we wait, maybe others will have a better insight

**Comment by louisjrdev** [comment]
@balazsorban44 I am not seeing anywhere in the example where getServerSideProps is making a call to a protected api endpoint with a getSession() call in it, am I missing something? I think that exact use case is where it is failing.

I did give the getServerSession a try, but that also did not work.

Thank you
